mono/metadata/sgen-tarjan-bridge.c

```diff
@@ -398,6 +439,13 @@ dyn_array_ptr_contains (DynPtrArray *da, void *x)
 }
 
 static gboolean
+match_colors_estimate (DynPtrArray *a, DynPtrArray *b)
+{
+	return dyn_array_ptr_size (a) == dyn_array_ptr_size (b);
+}
+
+
+static gboolean
 match_colors (DynPtrArray *a, DynPtrArray *b)
```

Comment:
Hashing is order independent. It does by adding the hashed version of each color pointer together.

Having said that. Adding elements when hashing is pretty bad as it has low avalanche properties, but it's the easy way around getting around the ordering problem.

The hash_perturb trick doesn't seem to help that much.

A simple test shows that the extra xoring introduces little to no impact into the actual hashcode, it cause some mild disturbance of low bits and none to the high bits. That together with the fact we do (hash & 0x7F) to compute the bucket translates into very little change. A simple test program shown it to a few times lead to streaks of repeating numbers. We're better off using a PRNG and stepping it on every collection.

This change will make debugging harder, as bugs repros won't be as stable.

The change to conservatively merge based on the hashcode makes a bit nervous, since we might end up merging too much and screwing up things. We should at least make it runtime toggleable in case we need to disable it for a problematic customer while we troubleshot.